xfrd-tcp.c

```diff
@@ -573,6 +662,50 @@ xfrd_tcp_open(struct xfrd_tcp_set* set, struct xfrd_tcp_pipeline* tp,
 	tp->tcp_r->fd = fd;
 	tp->tcp_w->fd = fd;
 
+	/* Check if an tls_auth name is configured which means we should try to
+	   establish an SSL connection */
+	if (zone->master->tls_auth_options &&
+		zone->master->tls_auth_options->auth_domain_name) {
+#ifdef HAVE_TLS_1_3
+		if (!setup_ssl(tp, set, zone->master->tls_auth_options->auth_domain_name)) {
+			log_msg(LOG_ERR, "xfrd: Cannot setup TLS on pipeline for %s to %s",
+					zone->apex_str, zone->master->ip_address_spec);
+			close(fd);
+			xfrd_set_refresh_now(zone);
+			return 0;
+		}
+		// TODO: There is a nasty case where the far end is listening on TCP 
+		// but not TLS. In that case the SSL_do_handshake function will loop, 
+		// returning SSL_ERROR_WANT_READ for the tcp_timeout (120s). This can 
+		// be handled various ways, not sure of the best one, need to discuss
+		// with Wouter...
+		int ret, err;
+		while (1) {
+			ERR_clear_error();
+			if ((ret = SSL_do_handshake(tp->ssl)) == 1) {
+				DEBUG(DEBUG_XFRD, 1, (LOG_INFO, "xfrd: TLS handshake sucessful"));
+				break;
+			}
+			err = SSL_get_error(tp->ssl, ret);
+			if (err != SSL_ERROR_WANT_READ && err != SSL_ERROR_WANT_WRITE) {
+				log_msg(LOG_ERR, "xfrd: TLS handshake failed for %s to %s with value: %d", 
+						zone->apex_str, zone->master->ip_address_spec, err);
+				close(fd);
+				xfrd_set_refresh_now(zone);
+				return 0;
+			}
+			/* else wants to be called again */
+		}
```

Comment:
So I cannot see a lot wrong with this.  You could add if(ret == 0) { close(fd); xfrd_set_refresh_now(zone); return 0; } before the 'else wants to be called again' line.  Then it would be fairly similar to other code that does not really have this issue.

Reply:
So this is indeed problematic, because it is spinning on the while loop (causing 100% cpu) and blocking the transfer daemon while doing that. This is resolved in commit 458519a (in PR #183 ) which moves the handshake into the async io based handling of the xfrd_tcp_pipe.

Reply:
That is the best fix! Makes the patch unproblematic for merge, in my opinion.